bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/handler/LivisiBridgeHandler.java

```diff
@@ -0,0 +1,936 @@
+
+import com.google.gson.Gson;
+
+/**
+ * The {@link LivisiBridgeHandler} is responsible for handling the LIVISI SmartHome controller including the connection
+ * to the LIVISI SmartHome backend for all communications with the LIVISI SmartHome {@link DeviceDTO}s.
+ * <p/>
+ * It implements the {@link AccessTokenRefreshListener} to handle updates of the oauth2 tokens and the
+ * {@link EventListener} to handle {@link EventDTO}s, that are received by the {@link LivisiWebSocket}.
+ * <p/>
+ * The {@link DeviceDTO}s are organized by the {@link DeviceStructureManager}, which is also responsible for the
+ * connection
+ * to the LIVISI SmartHome webservice via the {@link LivisiClient}.
+ *
+ * @author Oliver Kuhl - Initial contribution
+ * @author Hilbrand Bouwkamp - Refactored to use openHAB http and oauth2 libraries
+ * @author Sven Strohschein - Renamed from Innogy to Livisi
+ */
+@NonNullByDefault
+public class LivisiBridgeHandler extends BaseBridgeHandler
+        implements AccessTokenRefreshListener, EventListener, DeviceStatusListener {
+
+    private final Logger logger = LoggerFactory.getLogger(LivisiBridgeHandler.class);
+    private final Gson gson = new Gson();
+    private final Object lock = new Object();
+    private final Map<String, DeviceStatusListener> deviceStatusListeners;
+    private final OAuthFactory oAuthFactory;
+    private final HttpClient httpClient;
+
+    private @Nullable LivisiClient client;
+    private @Nullable LivisiWebSocket webSocket;
+    private @Nullable DeviceStructureManager deviceStructMan;
+    private @Nullable String bridgeId;
+    private @Nullable ScheduledFuture<?> reInitJob;
+    private @Nullable ScheduledFuture<?> bridgeRefreshJob;
+    private @NonNullByDefault({}) LivisiBridgeConfiguration bridgeConfiguration;
+    private @Nullable OAuthClientService oAuthService;
+    private String configVersion = "";
+
+    /**
+     * Constructs a new {@link LivisiBridgeHandler}.
+     *
+     * @param bridge Bridge thing to be used by this handler
+     * @param oAuthFactory Factory class to get OAuth2 service
+     * @param httpClient httpclient instance
+     */
+    public LivisiBridgeHandler(final Bridge bridge, final OAuthFactory oAuthFactory, final HttpClient httpClient) {
+        super(bridge);
+        this.oAuthFactory = oAuthFactory;
+        this.httpClient = httpClient;
+        deviceStatusListeners = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        // not needed
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(LivisiDeviceDiscoveryService.class);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing LIVISI SmartHome BridgeHandler...");
+        this.bridgeConfiguration = getConfigAs(LivisiBridgeConfiguration.class);
+        initializeClient();
```

Comment:
Status should be set before leaving `initialize()`:
```suggestion
        updateStatus(ThingStatus.UNKNOWN);
        initializeClient();
```

Reply:
At the moment the status is "INITIALIZING" for a few seconds until all data fetch requests are completed. When I change that with your suggestion, it is immediately "UNKNOWN" (the user can't see INITIALIZING because it is too fast set to UNKNOWN). Is that correct? The user sees "UNKNOWN" for a few seconds and after that "ONLINE" (or an error).

Reply:
That is indeed the correct behavior. The transition from **INITIALIZING** to **UNKNOWN** will be fast, because initialize() is required to return fast and not do blocking calls. Then usually after a few seconds, you will see next transition to **ONLINE** when scheduled jobs are triggered for the first time and network calls have succeeded.

Reply:
ok, thank you, than it is solved now (with the commit/push of today)